### DISCUSSION TITLE
Using Quectel voice over UAC

### DISCUSSION THREAD

**Discussion by IchthysMaranatha** [IchthysMaranatha/asterisk-chan-quectel#2] [Show and tell]
Quectel is slowly phasing out the audio serial port from its devices and moving on to voice through virtual USB sound card or UAC as can be seen here https://forums.quectel.com/t/eg25g-voice-over-usb/6113. 

Praise be to God! Have been able to copy over code from chan_alsa.c and integrate it with chan_quectel. There is now no need for bridging an Asterisk ALSA channel for using UAC.

UAC has been tested on my EC25-E and steps to use it are as follows

Enable UAC on the module, this is a one time setup which will persist through reboot
Issue command 
AT+QCFG="USBCFG" 

If UAC is not enabled, for my EC25 I'll receive the response
+QCFG: "usbcfg",0x2C7C,0x0125,1,1,1,1,1,0,0

Note that there should be 7 digits, either 1 or 0 after product code. The last digit is the UAC parameter, if there are less than seven parameters, then UAC is not available for the device. Now copy back the whole string and change last digit to 1 and issue new command

AT+QCFG="USBCFG",0x2C7C,0x0125,1,1,1,1,1,0,1

Now UAC will be enabled and aplay -L (or sudo aplay -L) should show you a new device for sound like 
hw:CARD=Android,DEV=0
    Android, USB Audio

Use new quectel.conf file, uncomment and set quec_uac=1, also uncomment alsadevice line (or use quectel.conf from uac folder)

**Comment by basncy** [comment]
Is it possible to migrate to entware for android? then we can turn a rooted Android phone itself into a gateway.

In my case, /dev/at_mdm0 is for AT commands.

**Comment by CpServiceSpb** [comment]
I hav one way sound only...
But withouch echo as with sound via NMEA.

**Reply by whiskerp** [reply]
same here

**Reply by huahuahuahuahuahuahuahua** [reply]
same, remote peer can not get my voice
when i enable UAC, i got  `hw:CARD=EC20CEHDLG` not `hw:CARD=Android`, I'm not sure if this difference is the cause.
asterisk record can record my voice, so i guess this is a UAC issue. 
change back to VoiceOverUSB is all ok, but low audio quality

**Reply by myleo1** [reply]
Do you still get only one-way audio even when configuring `hw:CARD=EC20CEHDLG` in the config? I'm encountering the same issue; after running `AT+QCFG="usbcfg",0x2C7C,0x0125,1,1,1,1,1,0,1`, I don't see an Android audio device, and I can't access ADB either.

**Comment by shmilyin** [comment]
my model is 
```Quectel
EC20F
Revision: EC20CEFAGR06A15M4G
```
i have got `hw:CARD=Android,DEV=0
Android, USB Audio` with aplay -L 
![image](https://github.com/user-attachments/assets/053ad696-ac94-418a-92cb-dda5bbd4e9cb)


and set qucetel0.conf to
 ```[quectel0]
;audio=/dev/ttyUSB1                       ; tty port for Audio, set as ttyUSB4 for Simcom if no other dev present
data=/dev/ttyUSB2                        ; tty port for AT commands;            no default value
quec_uac=1                              ; Uncomment line if using UAC mode
alsadev=hw:CARD=Android,DEV=0           ; Uncomment if using UAC, set device name or index as reqd
```
and 
 ```[quectel0]
audio=/dev/ttyUSB1                       ; tty port for Audio, set as ttyUSB4 for Simcom if no other dev present
data=/dev/ttyUSB2                        ; tty port for AT commands;            no default value
quec_uac=1                              ; Uncomment line if using UAC mode
alsadev=hw:CARD=Android,DEV=0           ; Uncomment if using UAC, set device name or index as reqd
```
but i got an error as 
```
-- [quectel0] Trying to connect on /dev/ttyUSB2...
[Feb 18 10:15:14] ERROR[1218]: chan_quectel.c:103 alsa_card_init: snd_pcm_open failed: No such device
[Feb 18 10:15:14] ERROR[1218]: chan_quectel.c:200 soundcard_init: Problem opening ALSA capture device hw:CARD=Android,DEV=0 
    -- [quectel0] Quectel has disconnected
    -- [quectel0] Quectel has connected, initializing...
[Feb 18 10:15:14] ERROR[1224]: at_queue.c:251 at_queue_run: [quectel0] Error write command 'AT' expected response 'OK' length 3, cancel
[Feb 18 10:15:14] ERROR[1224]: chan_quectel.c:617 do_monitor_phone: [quectel0] Error adding initialization commands to queue
```

this conf is work fine, but low audio quality
 ```[quectel0]
audio=/dev/ttyUSB1                       ; tty port for Audio, set as ttyUSB4 for Simcom if no other dev present
data=/dev/ttyUSB2                        ; tty port for AT commands;            no default value
;quec_uac=1                              ; Uncomment line if using UAC mode
;alsadev=hw:CARD=Android,DEV=0           ; Uncomment if using UAC, set device name or index as reqd
```
I have no idea what to do anymore.

**Reply by IchthysMaranatha** [reply]
Asterisk needs exclusive access to card, if using something like pulseaudio, it might not be able to take control and may throw such errors. If using non-root asterisk user, if it doesn't have full permission on the device, again may result in such errors.

**Reply by shmilyin** [reply]
Thank you very much for your reply. 
Below is my process, user group and device information.
```
root@debian11:~# ps -ef|grep asterisk
asterisk    1954       1  7 10:58 ?        00:00:01 /usr/sbin/asterisk -g -f -p -U asterisk -G asterisk
asterisk    1955    1954  0 10:58 ?        00:00:00 astcanary /var/run/asterisk/alt.asterisk.canary.tweet.tweet.tweet 1954
root        2035     595  0 10:58 pts/0    00:00:00 rasterisk rvvv
root        2039     839  0 10:59 pts/1    00:00:00 grep asterisk
root@debian11:~# 
root@debian11:~# ls -al /dev/ttyUSB*
crwxrwxrwx 1 asterisk dialout 188, 0 Feb 18 10:13 /dev/ttyUSB0
crwxrwxrwx 1 asterisk dialout 188, 1 Feb 18 10:13 /dev/ttyUSB1
crwxrwxrwx 1 asterisk dialout 188, 2 Feb 18 10:57 /dev/ttyUSB2
crwxrwxrwx 1 asterisk dialout 188, 3 Feb 18 10:13 /dev/ttyUSB3
root@debian11:~# 
root@debian11:~# 
root@debian11:~# groups asterisk 
asterisk : asterisk root dialout audio
```

**Reply by IchthysMaranatha** [reply]
It is a question of asterisk having full control over the alsa device, you might have to do a bit of search on how to achieve that and tinker around a bit, I may not be able to be of much help here. Also, remove all other tools that control audio devices.

**Reply by huahuahuahuahuahuahuahua** [reply]
> Thank you very much for your reply. Below is my process, user group and device information.
> 
> ```
> root@debian11:~# ps -ef|grep asterisk
> asterisk    1954       1  7 10:58 ?        00:00:01 /usr/sbin/asterisk -g -f -p -U asterisk -G asterisk
> asterisk    1955    1954  0 10:58 ?        00:00:00 astcanary /var/run/asterisk/alt.asterisk.canary.tweet.tweet.tweet 1954
> root        2035     595  0 10:58 pts/0    00:00:00 rasterisk rvvv
> root        2039     839  0 10:59 pts/1    00:00:00 grep asterisk
> root@debian11:~# 
> root@debian11:~# ls -al /dev/ttyUSB*
> crwxrwxrwx 1 asterisk dialout 188, 0 Feb 18 10:13 /dev/ttyUSB0
> crwxrwxrwx 1 asterisk dialout 188, 1 Feb 18 10:13 /dev/ttyUSB1
> crwxrwxrwx 1 asterisk dialout 188, 2 Feb 18 10:57 /dev/ttyUSB2
> crwxrwxrwx 1 asterisk dialout 188, 3 Feb 18 10:13 /dev/ttyUSB3
> root@debian11:~# 
> root@debian11:~# 
> root@debian11:~# groups asterisk 
> asterisk : asterisk root dialout audio
> ```


try 
```
bash -c "chown asterisk:asterisk -R /dev/snd/ && chown asterisk:asterisk -R /dev/ttyUSB*
```

**Reply by shmilyin** [reply]
```
root@debian11:~# ls -al /dev/snd/
total 0
drwxr-xr-x  4 root root      200 Feb 20 17:53 .
drwxr-xr-x 18 root root     3280 Feb 20 17:53 ..
drwxr-xr-x  2 root root       60 Feb 20 17:53 by-id
drwxr-xr-x  2 root root       80 Feb 20 17:53 by-path
crw-rw----  1 root audio 116,  5 Feb 20 17:53 controlC0
crw-rw----  1 root audio 116,  4 Feb 20 17:53 controlC1
crw-rw----  1 root audio 116,  3 Feb 20 17:53 pcmC1D0c
crw-rw----  1 root audio 116,  2 Feb 20 17:53 pcmC1D0p
crw-rw----  1 root audio 116,  1 Feb 20 17:53 seq
crw-rw----  1 root audio 116, 33 Feb 20 17:53 timer
```
after run `chown asterisk:asterisk -R /dev/snd/` was work for me. Thanks a lot.

**Reply by erwan2978** [reply]
Thanks to IchthysMaranatha to have presented here this solution.
After several days trying to restore without success,  the initial behavior with VOICE over the ttyUSB corresponding to (gps), I tried this and I get again output voice (from asterisk to the people calling my quectel "phone modem"), while during several day I got only incoming voice !!!
++++ This just as if my module quectel is resurrected ++++
To have no interference with ModemManager, I have choosen AT+QCFG="usbnet",1 (cdc_ether mode). So that no nead to get ModemManager active to have an active ethernet interface as WWAN
So I have something like this:
root@ismf38:~# **netstat -nr**
Kernel IP routing table
Destination      Gateway             Genmask         Flags   MSS Window  irtt Iface
0.0.0.0               192.168.200.2   0.0.0.0              UG        0 0                   0    enp2s0
0.0.0.0               192.168.225.1   0.0.0.0              UG        0 0                   0    enp0s22f2u1i4
192.168.200.0   0.0.0.0              255.255.255.0   U         0 0                    0   enp2s0
192.168.225.0   0.0.0.0              255.255.255.0   U         0 0                    0   enp0s22f2u1i4
==========================
As suggested by peoples here, astrerisk is in addtional :group audio, dialout
in /etc/rc.d/rc.local; each time I reboot, this script put also the good right on /run/lock, with:
chgrp dialout /run/lock; chmod g+w /run/lock.
I'm under Fedora 40, and I had to accommodate with selinux too, by doing all suggested actions by sealert, until no more warning popup from Selinux: and also but not sufficient:
~# semanage permissive -a asterisk_t  (add permissive selinux to all asterisk acction with asterisk_t   context; use -d like delete to remove, and semanage permissive -l to see all context in permissive mode: some are by default)
Without this permissive mode, other module like chan_mobile using bluetooth, fails, even when no more event are whown by sealert)
===========================
global result:
current people voice exchange is good (difficult to distinguish with standard voice over ttyUSB)
but the difference is in music on hold: less noise; sound is as clear than on a SIP trunk offered by operator like OVH over fixed Internet link (optical fiber FTTH). BUT in my case I hear some time something like jitters effect not smoothed.

**Reply by erwan2978** [reply]
excuse me for the magnifying glass effet over most of the comment. I do not understand how this is possible

**Reply by huahuahuahuahuahuahuahua** [reply]
@erwan2978 hello, I am using a EC20 with EC20CEHDLGR08A05M fw, I have tried many ways, but the peer could not hear my voice over UAC. 
Could you share what AT commonds you have configured?

**Reply by erwan2978** [reply]
To switch from audio via ttyUSB1n to audio with UACn I just executed the command suggested by IchthysMaranatha in this tutorial:
**AT+QCFG="USBCFG"**

and if you get an answer like for your  EC20: 
+QCFG: "usbcfg",0x2C7C,0x0120,1,1,1,1,1,0,0
(I have not an EC20n to control the parameters you got as default, so this is just an example with the device ID changed
The maintainer said:
you have to change the last digit is the UAC parameter, **if there are less than seven parameters, then UAC is not available for the device**. Now copy back the whole string and change last digit to 1 and issue new command

AT+QCFG="USBCFG",0x2C7C,0x0120,1,1,1,1,1,0,1

Did  you get an aswer OK, to this last command ?
And did you see in the listing of
root:~# **aplay -L**
a long listing like:

```
sysdefault:CARD=Android
    Android, USB Audio
    Default Audio Device
front:CARD=Android,DEV=0
    Android, USB Audio
    Front output / input
surround21:CARD=Android,DEV=0
    Android, USB Audio
    2.1 Surround output to Front and Subwoofer speakers
surround40:CARD=Android,DEV=0
    Android, USB Audio
    4.0 Surround output to Front and Rear speakers
surround41:CARD=Android,DEV=0
    Android, USB Audio
    4.1 Surround output to Front, Rear and Subwoofer speakers
surround50:CARD=Android,DEV=0
    Android, USB Audio
    5.0 Surround output to Front, Center and Rear speakers
surround51:CARD=Android,DEV=0
    Android, USB Audio
    5.1 Surround output to Front, Center, Rear and Subwoofer speakers
surround71:CARD=Android,DEV=0
    Android, USB Audio
    7.1 Surround output to Front, Center, Side, Rear and Woofer speakers
iec958:CARD=Android,DEV=0
    Android, USB Audio
    IEC958 (S/PDIF) Digital Audio Output
```
= = = =
Then at the end of  quectel.conf file, I have this;

```
data=/dev/ttyUSB2                        ; tty port for AT commands;            no default value
;audio=/dev/ttyUSB1                      ; tty port for Audio, set as ttyUSB4 for Simcom if no other dev present
quec_uac=1                              ; Uncomment line if using UAC mode
alsadev=hw:CARD=Android,DEV=0           ; Uncomment if using UAC, set device name or index as reqd
;imsi=                                   ; Fill in value, uncomment and comment out everything else - Not for UAC
;imei=                                   ; Fill in value, uncomment and comment out everything else - Not for UAC

```
So as you can see, no more usage of  /dev/ttyUSB for audio,  nor imei, nor ismi, as indicated by the maintainer,
Because some command are not taking into account with out a power of/power on, maybe I do it after this change. I do not remember. Just a remark of my context. Once again I have no EC20 to experiment. My asterisk is a VM in proxmox hypervisor, so power of/power on mean for me: shutdown the VM, and unplug, replug the EC25 which I have installed in a USB to miniPCIe case converter, in pupose to be able to do this. I do not use the internal miniPCIe slot of the  industrial mini PC, where Proxmox is Running, to avoid power of my Proxmox !

I just remember to you that I tried this because I was trapped in a situation like you, where on the /dev/ttyUSB0 as audio channel, no sound was outgoing on Mobile Channel. I could only hear the correspondant at the other end.

NOTICE also that in the comment of
 shmilyin on Feb 18, 
he gives 3 examples of entry in quectel.conf, and the second one is with both at the same time:
audio through ttyUSB1, and quec_uac, so this example is confusing indeed
```
audio=/dev/ttyUSB1                       ; tty port for Audio, set as ttyUSB4 for Simcom if no other dev present
data=/dev/ttyUSB2                        ; tty port for AT commands;            no default value
quec_uac=1                              ; Uncomment line if using UAC mode
alsadev=hw:CARD=Android,DEV=0           ; Uncomment if using UAC, set device name or index as reqd
```
So only his fisrt example is good, to use UAC.
= = = = 
So outside my comment here I do not how I could help you more, because I have no EC20 to test in your context

**Reply by erwan2978** [reply]
sorry I wanted to write above
_where on the /dev/ttyUSB1 as audio channel, no sound was outgoing on Mobile Channel._
and not               /dev/ttyUSB0

**Reply by huahuahuahuahuahuahuahua** [reply]
> To switch from audio via ttyUSB1n to audio with UACn I just executed the command suggested by IchthysMaranatha in this tutorial: **AT+QCFG="USBCFG"**
> 
> and if you get an answer like for your EC20: +QCFG: "usbcfg",0x2C7C,0x0120,1,1,1,1,1,0,0 (I have not an EC20n to control the parameters you got as default, so this is just an example with the device ID changed The maintainer said: you have to change the last digit is the UAC parameter, **if there are less than seven parameters, then UAC is not available for the device**. Now copy back the whole string and change last digit to 1 and issue new command
> 
> AT+QCFG="USBCFG",0x2C7C,0x0120,1,1,1,1,1,0,1
> 
> Did you get an aswer OK, to this last command ? And did you see in the listing of root:~# **aplay -L** a long listing like:
> 
> ```
> sysdefault:CARD=Android
>     Android, USB Audio
>     Default Audio Device
> front:CARD=Android,DEV=0
>     Android, USB Audio
>     Front output / input
> surround21:CARD=Android,DEV=0
>     Android, USB Audio
>     2.1 Surround output to Front and Subwoofer speakers
> surround40:CARD=Android,DEV=0
>     Android, USB Audio
>     4.0 Surround output to Front and Rear speakers
> surround41:CARD=Android,DEV=0
>     Android, USB Audio
>     4.1 Surround output to Front, Rear and Subwoofer speakers
> surround50:CARD=Android,DEV=0
>     Android, USB Audio
>     5.0 Surround output to Front, Center and Rear speakers
> surround51:CARD=Android,DEV=0
>     Android, USB Audio
>     5.1 Surround output to Front, Center, Rear and Subwoofer speakers
> surround71:CARD=Android,DEV=0
>     Android, USB Audio
>     7.1 Surround output to Front, Center, Side, Rear and Woofer speakers
> iec958:CARD=Android,DEV=0
>     Android, USB Audio
>     IEC958 (S/PDIF) Digital Audio Output
> ```
> 
> = = = = Then at the end of quectel.conf file, I have this;
> 
> ```
> data=/dev/ttyUSB2                        ; tty port for AT commands;            no default value
> ;audio=/dev/ttyUSB1                      ; tty port for Audio, set as ttyUSB4 for Simcom if no other dev present
> quec_uac=1                              ; Uncomment line if using UAC mode
> alsadev=hw:CARD=Android,DEV=0           ; Uncomment if using UAC, set device name or index as reqd
> ;imsi=                                   ; Fill in value, uncomment and comment out everything else - Not for UAC
> ;imei=                                   ; Fill in value, uncomment and comment out everything else - Not for UAC
> ```
> 
> So as you can see, no more usage of /dev/ttyUSB for audio, nor imei, nor ismi, as indicated by the maintainer, Because some command are not taking into account with out a power of/power on, maybe I do it after this change. I do not remember. Just a remark of my context. Once again I have no EC20 to experiment. My asterisk is a VM in proxmox hypervisor, so power of/power on mean for me: shutdown the VM, and unplug, replug the EC25 which I have installed in a USB to miniPCIe case converter, in pupose to be able to do this. I do not use the internal miniPCIe slot of the industrial mini PC, where Proxmox is Running, to avoid power of my Proxmox !
> 
> I just remember to you that I tried this because I was trapped in a situation like you, where on the /dev/ttyUSB0 as audio channel, no sound was outgoing on Mobile Channel. I could only hear the correspondant at the other end.
> 
> NOTICE also that in the comment of shmilyin on Feb 18, he gives 3 examples of entry in quectel.conf, and the second one is with both at the same time: audio through ttyUSB1, and quec_uac, so this example is confusing indeed
> 
> ```
> audio=/dev/ttyUSB1                       ; tty port for Audio, set as ttyUSB4 for Simcom if no other dev present
> data=/dev/ttyUSB2                        ; tty port for AT commands;            no default value
> quec_uac=1                              ; Uncomment line if using UAC mode
> alsadev=hw:CARD=Android,DEV=0           ; Uncomment if using UAC, set device name or index as reqd
> ```
> 
> So only his fisrt example is good, to use UAC. = = = = So outside my comment here I do not how I could help you more, because I have no EC20 to test in your context

Thank for your reply, I found that EC20 is not running an Android OS but running a small RTOS, so the alsa device is hw:CARD=EC20CEHDLG not hw:CARD=Android, this might be the difference. I decided to throw away EC20 and buy a EC25 to try

**Reply by abdofallah** [reply]
AT+QPCMV=1,2 needs to be executed to enable the audio on the UAC at start of the boot of the device as it doesn't save

**Reply by IchthysMaranatha** [reply]
That's done when receiving and placing calls, no need for it at boot

**Reply by abdofallah** [reply]
> That's done when receiving and placing calls, no need for it at boot

Understandable. A general question, have you noticed a ~0.5s delay in audio when using Quectel modules? shouldnt network gsm calls have 30ms delay only?

**Reply by IchthysMaranatha** [reply]
It's been some time since I've used it but do not recollect having a delay when using the quectel mini pcie module on the router with asterisk directly. Have had issues with delay when using something like ser2net+socat to route audio and com port to another device with asterisk.

**Comment by ryonwhyte** [comment]
Sharing what got me reliable two-way UAC voice on an EG25-G, in case it helps anyone here. (EG25-G, Asterisk 22, this fork, `quec_uac=1`.)

**1. Audio works on the first call, then goes silent until you restart Asterisk.**
The driver arms `AT+QPCMV=0;+QPCMV=1,2;A` at answer but never tears the PCM session down on hangup, so it stays armed and the next call's `QPCMV` toggle never re-inits cleanly.

You can confirm it with no code. After a wedged call, run `asterisk -rx 'quectel cmd <dev> AT+QPCMV=0'`, wait about 5 seconds, and the next call works.

The fix is to enqueue `QPCMV=0` on hangup, in the UAC branch of `disactivate_call()` (`channel.c`):
```c
at_enqueue_user_cmd(cpvt, "AT+QPCMV=0");
```
That one change is what made back-to-back calls work for me.

**2. Garbled, one-way, or muted downlink? Try `AT+QAUDMOD=3`.**
The EG25-G's default handset DSP (echo canceller plus noise suppressor) mangles the raw UAC stream. You get screeching, or a downlink the remote can't hear even though your own capture records fine. `AT+QAUDMOD=3` turns all the processing off and gives you clean PCM both ways. It saves to NVM, but I re-assert it on boot to be safe.

One more thing on that: stock firmware `30.200` muted my downlink entirely, and `A0.304` fixed it. So check your firmware if `QAUDMOD=3` alone isn't enough.

**3. If you run separate capture and playback threads,** join both of them before you call `snd_pcm_close()` on hangup. Closing the playback handle while its thread is still inside `snd_pcm_writei()` corrupts the gadget.

**4. `snd_pcm_open failed: No such device`** is usually an `alsadev` mismatch, not a hardware fault. The driver appends `cap` and `play` to your `alsadev` value, so it has to resolve to real PCMs (the names in your `asound.conf`, or a real card), not a bare device path. Also don't set both `audio=` and `quec_uac=1`.

### Update
I flashed the latest firmware and no longer needed any of tyese things. You can obtain directly from quectel.

**Reply by lucsansag** [reply]
> Sharing what got me reliable two-way UAC voice on an EG25-G, in case it helps anyone here. (EG25-G, Asterisk 22, this fork, `quec_uac=1`.)
> 
> **1. Audio works on the first call, then goes silent until you restart Asterisk.** The driver arms `AT+QPCMV=0;+QPCMV=1,2;A` at answer but never tears the PCM session down on hangup, so it stays armed and the next call's `QPCMV` toggle never re-inits cleanly.
> 
> You can confirm it with no code. After a wedged call, run `asterisk -rx 'quectel cmd <dev> AT+QPCMV=0'`, wait about 5 seconds, and the next call works.
> 
> The fix is to enqueue `QPCMV=0` on hangup, in the UAC branch of `disactivate_call()` (`channel.c`):
> 
> ```c
> at_enqueue_user_cmd(cpvt, "AT+QPCMV=0");
> ```
> 
> That one change is what made back-to-back calls work for me.
> 
> **2. Garbled, one-way, or muted downlink? Try `AT+QAUDMOD=3`.** The EG25-G's default handset DSP (echo canceller plus noise suppressor) mangles the raw UAC stream. You get screeching, or a downlink the remote can't hear even though your own capture records fine. `AT+QAUDMOD=3` turns all the processing off and gives you clean PCM both ways. It saves to NVM, but I re-assert it on boot to be safe.
> 
> One more thing on that: stock firmware `30.200` muted my downlink entirely, and `A0.304` fixed it. So check your firmware if `QAUDMOD=3` alone isn't enough.
> 
> **3. If you run separate capture and playback threads,** join both of them before you call `snd_pcm_close()` on hangup. Closing the playback handle while its thread is still inside `snd_pcm_writei()` corrupts the gadget.
> 
> **4. `snd_pcm_open failed: No such device`** is usually an `alsadev` mismatch, not a hardware fault. The driver appends `cap` and `play` to your `alsadev` value, so it has to resolve to real PCMs (the names in your `asound.conf`, or a real card), not a bare device path. Also don't set both `audio=` and `quec_uac=1`.

@ryonwhyte I was thinking of trying this driver and I came across your comment. Regarding 1., I see that branch currently contains only
```c
snd_pcm_drop(pvt->icard); 
```
,does it matter if the statement you wrote is added before or after that one? Thanks in advance!

**Reply by ryonwhyte** [reply]
hmmm. I got an updated firmware, flashed the modems and everything worked. I then removed all of these little workarounds one at a time and things are fine. I should have updated here. What firmware version are you running?

**Reply by lucsansag** [reply]
I have an EC25 running EC25EUXGAR08A17M1G_20.200.20.200, I haven't updated the firmware or attempted using this driver yet.

**Reply by ryonwhyte** [reply]
Oooh. There are probably 2-4 newer releases - This was the latest I see referenced EC25EUXGAR08A19M1G_A0.302.A0.302 but I think there could also be a ....304 based on how the version numbers look. 

Before you do anything I would recommend updating to the latest firmware. It really did solve all of my issues and could have saved me so much time.
